File: test/language/builtin_types/cpp/DynamicBitFieldLengthBoundsTest.cpp
#include "builtin_types/dynamic_bitfield_length_bounds/Container.h"
#include "gtest/gtest.h"
#include "zserio/SerializeUtil.h"

namespace builtin_types
{
namespace dynamic_bitfield_length_bounds
{

using allocator_type = Container::allocator_type;
using BitBuffer = zserio::BasicBitBuffer<zserio::RebindAlloc<allocator_type, uint8_t>>;

class DynamicBitFieldLengthBoundsTest : public ::testing::Test
{
protected:
    void writeContainer(const Container container)
    {
        zserio::BitStreamWriter writer(bitBuffer);
        writer.writeBits(container.getUnsignedBitLength(), 4);
        if (container.getUnsignedBitLength() == 0)
        {
            return;
        }
        writer.writeBits(container.getUnsignedValue(), container.getUnsignedBitLength());
        writer.writeBits(container.getUnsignedBigBitLength(), 8);
        if (container.getUnsignedBigBitLength() == 0 || container.getUnsignedBigBitLength() > 64)
        {
            return;
        }
        writer.writeBits64(container.getUnsignedBigValue(), container.getUnsignedBigBitLength());
        writer.writeBits64(container.getSignedBitLength(), 64);
        if (container.getSignedBitLength() == 0 || container.getSignedBitLength() > 64)
        {
            return;
        }
        writer.writeSignedBits64(
                container.getSignedValue(), static_cast<uint8_t>(container.getSignedBitLength()));
    }

    static constexpr uint8_t UNSIGNED_BIT_LENGTH = 15;
    static constexpr uint16_t UNSIGNED_VALUE = (1U << UNSIGNED_BIT_LENGTH) - 1;
    static constexpr uint8_t UNSIGNED_BIG_BIT_LENGTH = 13;
    static constexpr uint64_t UNSIGNED_BIG_VALUE = (1U << UNSIGNED_BIG_BIT_LENGTH) - 1;
    static constexpr uint64_t SIGNED_BIT_LENGTH = 7;
    static constexpr int64_t SIGNED_VALUE = -static_cast<int64_t>(1U << (SIGNED_BIT_LENGTH - 1));

    BitBuffer bitBuffer = BitBuffer(1024);
};

TEST_F(DynamicBitFieldLengthBoundsTest, writeRead)
{
    Container container = {UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH, UNSIGNED_BIG_VALUE,
            SIGNED_BIT_LENGTH, SIGNED_VALUE};

    bitBuffer = zserio::serialize(container);
    const Container readContainer = zserio::deserialize<Container>(bitBuffer);
    ASSERT_EQ(container, readContainer);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    Container readContainer2(reader);
    ASSERT_EQ(container, readContainer2);
}

TEST_F(DynamicBitFieldLengthBoundsTest, unsignedBitLengthZero)
{
    Container container = {
            0, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH, UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE};

    ASSERT_THROW(zserio::serialize(container), zserio::CppRuntimeException);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    ASSERT_THROW(Container readContainer(reader), zserio::CppRuntimeException);
}

TEST_F(DynamicBitFieldLengthBoundsTest, unsignedBitLengthZeroValueZero)
{
    Container container = {0, 0, UNSIGNED_BIG_BIT_LENGTH, UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE};

    ASSERT_THROW(zserio::serialize(container), zserio::CppRuntimeException);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    ASSERT_THROW(Container readContainer(reader), zserio::CppRuntimeException);
}

TEST_F(DynamicBitFieldLengthBoundsTest, unsignedBigBitLengthZero)
{
    Container container = {
            UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, 0, UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE};

    ASSERT_THROW(zserio::serialize(container), zserio::CppRuntimeException);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    ASSERT_THROW(Container readContainer(reader), zserio::CppRuntimeException);
}

TEST_F(DynamicBitFieldLengthBoundsTest, unsignedBigBitLengthZeroValueZero)
{
    Container container = {UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, 0, 0, SIGNED_BIT_LENGTH, SIGNED_VALUE};

    ASSERT_THROW(zserio::serialize(container), zserio::CppRuntimeException);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    ASSERT_THROW(Container readContainer(reader), zserio::CppRuntimeException);
}

TEST_F(DynamicBitFieldLengthBoundsTest, unsignedBigBitLengthOverMax)
{
    Container container = {
            UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, 65, UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE};

    ASSERT_THROW(zserio::serialize(container), zserio::CppRuntimeException);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    ASSERT_THROW(Container readContainer(reader), zserio::CppRuntimeException);
}

TEST_F(DynamicBitFieldLengthBoundsTest, signedBitLengthZero)
{
    Container container = {
            UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH, UNSIGNED_BIG_VALUE, 0, SIGNED_VALUE};

    ASSERT_THROW(zserio::serialize(container), zserio::CppRuntimeException);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    ASSERT_THROW(Container readContainer(reader), zserio::CppRuntimeException);
}

TEST_F(DynamicBitFieldLengthBoundsTest, signedBitLengthZeroValueZero)
{
    Container container = {
            UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH, UNSIGNED_BIG_VALUE, 0, 0};

    ASSERT_THROW(zserio::serialize(container), zserio::CppRuntimeException);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    ASSERT_THROW(Container readContainer(reader), zserio::CppRuntimeException);
}

TEST_F(DynamicBitFieldLengthBoundsTest, signedBitLengthOverMax)
{
    Container container = {
            UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH, UNSIGNED_BIG_VALUE, 65, SIGNED_VALUE};

    ASSERT_THROW(zserio::serialize(container), zserio::CppRuntimeException);

    writeContainer(container);
    zserio::BitStreamReader reader(bitBuffer);
    ASSERT_THROW(Container readContainer(reader), zserio::CppRuntimeException);
}

} // namespace dynamic_bitfield_length_bounds
} // namespace builtin_types

File: test/extensions/language/builtin_types/java/builtin_types/DynamicBitFieldLengthBoundsTest.java
package builtin_types;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.math.BigInteger;

import org.junit.jupiter.api.Test;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.SerializeUtil;

import builtin_types.dynamic_bitfield_length_bounds.Container;

public class DynamicBitFieldLengthBoundsTest
{
    @Test
    public void writeRead() throws IOException
    {
        final Container container = new Container(UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH,
                UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE);

        final BitBuffer bitBuffer = SerializeUtil.serialize(container);
        final Container readContainer = SerializeUtil.deserialize(Container.class, bitBuffer);

        assertEquals(container, readContainer);
    }

    @Test
    public void unsignedBitLengthZero() throws IOException
    {
        final Container container = new Container((byte)0, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH,
                UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE);

        assertThrows(IllegalArgumentException.class, () -> SerializeUtil.serialize(container));

        final byte[] bytes = writeContainer(container);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        assertThrows(IllegalArgumentException.class, () -> new Container(reader));
    }

    @Test
    public void unsignedBitLengthZeroValueZero() throws IOException
    {
        final Container container = new Container((byte)0, (short)0, UNSIGNED_BIG_BIT_LENGTH,
                UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE);

        assertThrows(IllegalArgumentException.class, () -> SerializeUtil.serialize(container));

        final byte[] bytes = writeContainer(container);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        assertThrows(IllegalArgumentException.class, () -> new Container(reader));
    }

    @Test
    public void unsignedBigBitLengthZero() throws IOException
    {
        final Container container = new Container(UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, (short)0,
                UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE);

        assertThrows(IllegalArgumentException.class, () -> SerializeUtil.serialize(container));

        final byte[] bytes = writeContainer(container);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        assertThrows(IllegalArgumentException.class, () -> new Container(reader));
    }

    @Test
    public void unsignedBigBitLengthZeroValueZero() throws IOException
    {
        final Container container = new Container(UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, (short)0,
                BigInteger.ZERO, SIGNED_BIT_LENGTH, SIGNED_VALUE);

        assertThrows(IllegalArgumentException.class, () -> SerializeUtil.serialize(container));

        final byte[] bytes = writeContainer(container);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        assertThrows(IllegalArgumentException.class, () -> new Container(reader));
    }

    @Test
    public void unsignedBigBitLengthOverMax() throws IOException
    {
        final Container container = new Container(UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, (short)65,
                UNSIGNED_BIG_VALUE, SIGNED_BIT_LENGTH, SIGNED_VALUE);

        assertThrows(IllegalArgumentException.class, () -> SerializeUtil.serialize(container));

        final byte[] bytes = writeContainer(container);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        assertThrows(IllegalArgumentException.class, () -> new Container(reader));
    }

    @Test
    public void signedBitLengthZero() throws IOException
    {
        final Container container = new Container(UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH,
                UNSIGNED_BIG_VALUE, BigInteger.valueOf(0), SIGNED_VALUE);

        assertThrows(IllegalArgumentException.class, () -> SerializeUtil.serialize(container));

        final byte[] bytes = writeContainer(container);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        assertThrows(IllegalArgumentException.class, () -> new Container(reader));
    }

    @Test
    public void signedBitLengthZeroValueZero() throws IOException
    {
        final Container container = new Container(UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH,
                UNSIGNED_BIG_VALUE, BigInteger.valueOf(0), (long)0);

        assertThrows(IllegalArgumentException.class, () -> SerializeUtil.serialize(container));

        final byte[] bytes = writeContainer(container);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        assertThrows(IllegalArgumentException.class, () -> new Container(reader));
    }

    @Test
    public void signedBitLengthOverMax() throws IOException
    {
        final Container container = new Container(UNSIGNED_BIT_LENGTH, UNSIGNED_VALUE, UNSIGNED_BIG_BIT_LENGTH,
                UNSIGNED_BIG_VALUE, BigInteger.valueOf(65), SIGNED_VALUE);

        assertThrows(IllegalArgumentException.class, () -> SerializeUtil.serialize(container));

        final byte[] bytes = writeContainer(container);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        assertThrows(IllegalArgumentException.class, () -> new Container(reader));
    }

    byte[] writeContainer(Container container) throws IOException
    {
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();

        writer.writeBits(container.getUnsignedBitLength(), 4);
        if (container.getUnsignedBitLength() < 1)
            return writer.toByteArray();
        writer.writeBits(container.getUnsignedValue(), container.getUnsignedBitLength());
        writer.writeBits(container.getUnsignedBigBitLength(), 8);
        if (container.getUnsignedBigBitLength() < 1 || container.getUnsignedBigBitLength() > 64)
            return writer.toByteArray();
        writer.writeBigInteger(container.getUnsignedBigValue(), container.getUnsignedBigBitLength());
        writer.writeBigInteger(container.getSignedBitLength(), 64);
        if (container.getSignedBitLength().compareTo(BigInteger.ONE) == -1 ||
                container.getSignedBitLength().compareTo(BigInteger.valueOf(64)) == 1)
            return writer.toByteArray();
        writer.writeBits(container.getSignedValue(), container.getSignedBitLength().intValue());

        return writer.toByteArray();
    }

    private static final byte UNSIGNED_BIT_LENGTH = 15;
    private static final short UNSIGNED_VALUE = (1 << 15) - 1;
    private static final short UNSIGNED_BIG_BIT_LENGTH = 13;
    private static final BigInteger UNSIGNED_BIG_VALUE =
            BigInteger.ONE.shiftLeft(UNSIGNED_BIG_BIT_LENGTH).subtract(BigInteger.ONE);
    private static final BigInteger SIGNED_BIT_LENGTH = BigInteger.valueOf(7);
    private static final long SIGNED_VALUE = -(1L << (SIGNED_BIT_LENGTH.intValue() - 1));
}
